include/fiction/algorithms/simulation/sidb/automatic_exhaustive_gate_designer.hpp

```diff
@@ -0,0 +1,273 @@
+    /**
+     * Stores various combinations of SiDB indices, representing the cell positions
+     * within the canvas. These combinations are used to distribute a specified number of SiDBs
+     * across the canvas (all possible distributions are covered), typically from top to bottom and left to right.
+     */
+    std::vector<std::vector<std::size_t>> all_combinations{};
+    /**
+     * Number of canvas cells.
+     */
+    std::size_t number_of_canvas_cells{};
+    /**
+     * All found SiDB cell-level layouts that fulfill the given truth table.
+     */
+    std::vector<Lyt> all_found_gate_layouts{};
+
+    /**
+     * Initialize the *Automatic Exhaustive Gate Designer*.
+     *
+     * This function initializes various internal data structures and parameters required for
+     * the automatic exhaustive gate designer. It calculates the set of SiDBs within the specified
+     * canvas area, determines the total number of canvas cells, and prepares containers for
+     * storing combinations and found layouts.
+     */
+    void initialize() noexcept
+    {
+        all_sidbs_in_cavas     = all_sidbs_in_spanned_area(parameter.canvas.first, parameter.canvas.second);
+        number_of_canvas_cells = all_sidbs_in_cavas.size();
+    }
+
+    /**
+     * Calculates all possible combinations of distributing the given number of SiDBs within a canvas
+     * based on the provided parameters. It generates combinations of SiDB indices (representing the cell position in
+     * the canvas from top to bottom and left to right) and stores them in the `all_combinations` vector. The number of
+     * SiDBs in each combination is determined by `parameter.number_of_sidbs`.
+     *
+     * @note The function uses a combinatorial algorithm to generate the combinations.
+     *
+     */
+    void determine_all_combinations_of_given_sidbs_in_canvas()
+    {
+        all_combinations = {};
+        all_combinations.reserve(binomial_coefficient(number_of_canvas_cells, parameter.number_of_sidbs));
+        std::vector<std::size_t> numbers(number_of_canvas_cells);
+        std::iota(numbers.begin(), numbers.end(), 0);
+        combinations::for_each_combination(
+            numbers.begin(),
+            numbers.begin() + static_cast<std::vector<std::size_t>::difference_type>(parameter.number_of_sidbs),
+            numbers.end(),
+            [this](const auto begin, const auto end)
+            {
+                std::vector<std::size_t> combination{};
+                combination.reserve(parameter.number_of_sidbs);
+                for (auto it = begin; it != end; ++it)
+                {
+                    combination.push_back(static_cast<std::size_t>(*it));
+                }
+                all_combinations.push_back(combination);
+                return false;  // keep looping
+            });
+    }
+
+    [[maybe_unused]] bool are_sidbs_to_close(const std::vector<std::size_t>& cell_indices) noexcept
```

Comment:
ah, sorry! I just noticed that the docstring and some comments are missing here. I will fix it.